src/ability/msg.rs

```diff
@@ -0,0 +1,183 @@
+use crate::{ability::traits::Command, prove::TryProve};
+use libipld_core::ipld::Ipld;
+use std::{collections::BTreeMap, str::FromStr};
+use url::Url;
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Msg {
+    to: Url,
+    from: Url,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct MsgSend {
+    to: Url,
+    from: Url,
+    message: String,
+}
+
+// TODO is the to or from often also the subject? Shoudl that be accounted for?
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct MsgReceive {
+    to: Url,
+    from: Url,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct MsgReceiveBuilder {
+    to: Option<Url>,
+    from: Option<Url>,
+}
+
+impl From<MsgReceive> for MsgReceiveBuilder {
+    fn from(msg: MsgReceive) -> Self {
+        Self {
+            to: Some(msg.to),
+            from: Some(msg.from),
+        }
+    }
+}
+
+impl TryFrom<MsgReceiveBuilder> for MsgReceive {
+    type Error = MsgReceiveBuilder;
+
+    fn try_from(builder: MsgReceiveBuilder) -> Result<Self, MsgReceiveBuilder> {
+        // FIXME
+        if let (Some(to), Some(from)) = (builder.clone().to, builder.clone().from) {
+            Ok(Self { to, from })
+        } else {
+            Err(builder.clone()) // FIXME
+        }
+    }
+}
+
+impl From<MsgReceive> for Ipld {
+    fn from(msg: MsgReceive) -> Self {
+        let mut map = BTreeMap::new();
+        map.insert("to".into(), msg.to.to_string().into());
+        map.insert("from".into(), msg.from.to_string().into());
+        map.into()
+    }
+}
+
+impl TryFrom<&Ipld> for MsgReceiveBuilder {
```

Comment:
Rather than manually writing this implementation, if you'd like, you can probably use `libipld_core` with the `serde-codec` feature to derive `Deserialize` on `MsgReceive`, and then use `IntoDeserializer ` to do `MsgReceive::deserialize(some_ipld.into_deserializer())`.

Something like this:

```rust
use serde::de::{Deserialize, IntoDeserializer};
use serde_derive::Deserialize;

#[derive(Debug, Clone, PartialEq, Eq, Deserialize)]
pub struct MsgReceive {
    to: Url,
    from: Url,
}

impl TryFrom<&Ipld> for MsgReceiveBuilder {
    type Error = ();

    fn try_from(ipld: &Ipld) -> Result<Self, ()> {
        MsgReceive::deserialize(some_ipld.into_deserializer()).map_err(|_| ())
    }
}
```

Reply:
> `the trait Deserialize<'_> is not implemented for Url`

Womp womp

But good tip in general! I guess I can newtype `Url`, but it's some code either way

Reply:
@expede are you importing the url lib w/ the serde feature?

Reply:
@zeeshanlakhani good point! Ugh, I really dislike how invisible those are in tooling.

Reply:
What I learned:

```toml
libipld-core = { version = "0.16", features = ["serde-codec"] }
```

```rust
use libipld_core::{ipld::Ipld, serde as ipld_serde};

// ...

#[derive(Debug, Clone, PartialEq, Eq, Deserialize)]
pub struct MsgReceive {
    to: Url,
    from: Url,
}

#[derive(Debug, Clone, PartialEq, Eq, Deserialize)]
pub struct MsgReceiveBuilder {
    to: Option<Url>,
    from: Option<Url>,
}

// ...

impl TryFrom<Ipld> for MsgReceiveBuilder {
    type Error = ();

    fn try_from(ipld: Ipld) -> Result<Self, ()> {
        ipld_serde::from_ipld(ipld).map_err(|_| ())
    }
}
```